### PR TITLE
feat: support account name management from snap

### PR DESCRIPTION
**Pull request by khanti42**
# PR Summary: Account Naming in Snap

### 1. Account Naming
- Introduces an `accountName` field for accounts.
- Default names follow the pattern: `"Account {index}"` (e.g., `Account 1`, `Account 2`).
- Names are assigned during:
  - Account creation
  - Account recovery
- Users can rename accounts via the new `SetAccountName` RPC method.

### 2. New RPC Methods
- **`SetAccountName`**: Allows renaming an account.

### **Account Management Enhancements**
- Modified `AccountStateManager` to store and update `accountName`.
- Introduced `getNextIndex()` in `AccountService` to fetch the next account index, used in the frontend to display the default Name.
- Updated `createAccount`, `recoverAccounts`, and `addAccount` to handle account names.

### **RPC Handling**
- Added new RPC handlers:
  - `SetAccountNameRpc` (for renaming accounts)
- Updated `onRpcRequest` to support these methods.

### **Testing**
- Added test cases for:
  - `setAccountName`
  - `getNextAccountIndex`
- Updated existing tests to validate account name assignments.

- Removed unnecessary whitespace in `config.ts`.
- Updated `superstruct.ts` to validate `accountName`.

BEGIN_COMMIT_OVERRIDE

feat: Support account name management in Snap

END_COMMIT_OVERRIDE

### CODE REVIEW

File: packages/starknet-snap/src/rpcs/set-account-name.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    const accountJsonData = await accMgt.withTransaction(async (state) => {
+      await accMgt.setAccountName({
+        address,
+        chainId,
+        accountName,
+      });
+      return await accMgt.getCurrentAccount({ chainId }, state);
+    });
+
+    return accountJsonData as SetAccountNameResponse;
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.ts:60]: do you have a use case to return the account

it is different, 
ToggleAccount need to return the latest current account, but it seem account name doesnt required?

File: packages/starknet-snap/src/rpcs/set-account-name.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    const { address, chainId } = this.account;
+
+    const accountJsonData = await accMgt.withTransaction(async (state) => {
+      await accMgt.setAccountName({
+        address,
+        chainId,
+        accountName,
+      });
+      return await accMgt.getCurrentAccount({ chainId }, state);
+    });
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.ts:58]: you dont need an extra method to set the name
```suggestion
    await accMgt.upsertAccount({
       ...await account.serialize(),
       accountName
    })
```

File: packages/starknet-snap/src/state/account-state-manager.ts [hunk 3]
```diff
@@ -228,6 +229,46 @@ export class AccountStateManager extends StateManager<AccContract> {
...
+  /**
+   * Sets the name of a account by address and chain ID.
+   *
+   * @param params - The parameters for toggle the account visibility.
+   * @param params.address - The address of the account.
+   * @param params.chainId - The chain ID of the account.
+   * @param params.accountName - The name of the account.
+   * @throws {StateManagerError} If the account to be removed does not exist.
+   */
+  async setAccountName({
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/state/account-state-manager.ts:241]: we dont need this method, we have upsert

File: packages/starknet-snap/src/utils/superstruct.ts [hunk 4]
```diff
@@ -437,4 +437,5 @@ export const AccountStruct = object({
   upgradeRequired: boolean(),
   deployRequired: boolean(),
   visibility: optional(boolean()),
+  accountName: string(),
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/utils/superstruct.ts:440]:
```suggestion
  accountName: optional(string()),
```

File: packages/starknet-snap/src/wallet/account/account.ts [hunk 5]
```diff
@@ -1,11 +1,10 @@
 import type { CairoVersion } from 'starknet';
 
-import type { AccContract } from '../../types/snapState';
+import type { AccContract, AccountMetaData } from '../../types/snapState';
 import type { CairoAccountContract } from './contract';
 
-export type AccountMetaData = { visibility?: boolean };
-
 export const DefaultAccountMetaData: AccountMetaData = {
+  accountName: 'Account 1',
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/wallet/account/account.ts:7]: change to Utils
```suggestion
  accountName: getDefaultAccountName(0)
```

File: packages/starknet-snap/src/state/account-state-manager.test.ts [hunk 6]
```diff
@@ -160,6 +160,21 @@ describe('AccountStateManager', () => {
     });
   });
 
+  it('updates the account name', async () => {
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/state/account-state-manager.test.ts:163]: try not to put a it without describe

File: packages/starknet-snap/src/wallet/account/account.ts [hunk 7]
```diff
@@ -63,13 +62,18 @@ export class Account {
...
-  #jsonDataToMetaData(jsonData?: AccContract): void {
+  #jsonDataToMetaData(jsonData?: Partial<AccContract>): void {
     if (!jsonData) {
       return;
     }
     if (jsonData.visibility !== undefined) {
       this.metadata.visibility = jsonData.visibility;
     }
+    if (jsonData.accountName === undefined) {
+      this.metadata.accountName = `Account ${this.hdIndex + 1}`;
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/wallet/account/account.ts:73]:
```suggestion
      this.metadata.accountName = getDefaultAccountName(0)
```

File: packages/starknet-snap/src/wallet/account/service.ts [hunk 8]
```diff
@@ -37,7 +43,7 @@ export class AccountService {
    */
   async deriveAccountByIndex(
     index?: number,
-    jsonData?: AccContract,
+    jsonData?: AccountMetaData,
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/wallet/account/service.ts:46]: in terms of typing, i feel it is much better than `AccContract`

but since deriveAccountByAddress is passing the accountFromState into this method

for clear coding, may be we should change deriveAccountByAddress to only provide AccountMetaData rather than the whole account obj

File: packages/starknet-snap/test/constants.test.ts [hunk 9]
```diff
@@ -54,6 +56,7 @@ export const account3: AccContract = {
   addressSalt:
     '0x0797efd8e3971dfca4f928485860896201320ce2997ca4789d9343204219599d',
   addressIndex: 2,
+  accountName: `Account ${2 + 1}`,
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/constants.test.ts:59]: if the accountName is optional , we dont need those

File: packages/starknet-snap/src/state/account-state-manager.test.ts (hunk 6 above)
Comment (khanti42) [reply packages/starknet-snap/src/state/account-state-manager.test.ts]: removed

File: packages/starknet-snap/src/rpcs/set-account-name.ts (hunk 1 above)
Comment (khanti42) [reply packages/starknet-snap/src/rpcs/set-account-name.ts]: No use case for now but I kept it, let me know if you think its not ok

File: packages/starknet-snap/src/utils/account.test.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { getDefaultAccountName } from './account';
+
+describe('getDefaultAccountName', () => {
+  it.each([
+    [undefined, 'Account 1'], // Default case
+    [0, 'Account 1'],
+    [1, 'Account 2'],
+    [5, 'Account 6'],
+    [999, 'Account 1000'],
+    [-1, 'Account 0'], // Edge case: negative number
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/utils/account.test.ts:10]: nit:
can we not accept negative / float number

File: packages/starknet-snap/src/utils/superstruct.ts [hunk 11]
```diff
@@ -219,6 +219,17 @@ export const StarkNameStruct = refine(
   },
 );
 
+export const AccountNameStruct = refine(
+  string(),
+  'AccountNameStruct',
+  (value: string) => {
+    if (value.length < 20) {
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/utils/superstruct.ts:226]: min length also require i guess

File: packages/starknet-snap/src/rpcs/set-account-name.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    const { accountName, chainId } = params;
+
+    const accMgt = new AccountStateManager();
+
+    await accMgt.updateAccountByAddress({
+      ...(await this.account.serialize()),
+      accountName,
+    });
+
+    return (await accMgt.getCurrentAccount({
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.ts:59]: i think it better to return the request account, rather than the current account

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]
Almost prefect, just 3 comments for enhancement

File: packages/starknet-snap/src/rpcs/set-account-name.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+   */
+  protected async handleRequest(
+    params: SetAccountNameParams,
+  ): Promise<SetAccountNameResponse> {
+    const { accountName } = params;
+
+    const accMgt = new AccountStateManager();
+
+    await accMgt.updateAccountByAddress({
+      ...(await this.account.serialize()),
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.ts:55]: lets not duplicate this.account.serialize() twice

File: packages/starknet-snap/src/utils/superstruct.ts [hunk 14]
```diff
@@ -219,6 +219,17 @@ export const StarkNameStruct = refine(
   },
 );
 
+export const AccountNameStruct = refine(
+  string(),
+  'AccountNameStruct',
+  (value: string) => {
+    if (value.length > 1 && value.length < 20) {
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/utils/superstruct.ts:226]: so the min length is 2? and max length is 19?

File: packages/starknet-snap/src/rpcs/set-account-name.test.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+describe('SetAccountName', () => {
+  const network = STARKNET_SEPOLIA_TESTNET_NETWORK;
+
+  const setupSetAccountNameTest = async (accountName = 'My Account Name') => {
+    const { accountObj: account } = await createAccountObject(network);
+
+    mockAccountContractReader({});
+
+    const { getAccountSpy, getCurrentAccountSpy } = mockAccountStateManager();
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.test.ts:25]: since u dont need getCurrentAccountSpy anymore, remove those no use spy and mock

File: packages/starknet-snap/src/rpcs/set-account-name.test.ts [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        address,
+        addressSalt,
+        chainId,
+        hdIndex,
+        cairoVersion,
+        publicKey,
+      },
+      request,
+      updateAccountByAddressSpy,
+    } = await setupSetAccountNameTest(accountName);
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/set-account-name.test.ts:63]: it does not seem there is a test for checking if the metadata of name is given or not (you can do the test in service.ts)
test case:
name is undefined -> expect default name
name is defined -> expect defined name

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]
few more comments

**Review by stanleyyconsensys — APPROVED** [review]
